### DISCUSSION TITLE
JMESPath Community 1.1

### DISCUSSION THREAD

**Discussion by springcomp** [jmespath-community/jmespath.spec#97] [Announcements]
**JMESPath Community 1.1.0** implements most of those ideas.

After working for many months on [various ideas](https://github.com/jmespath-community/jmespath.spec/discussions/38) we think that we have come up with a reasonable set of useful improvements to the JMESPath Community specification.

<img src="https://user-images.githubusercontent.com/1768820/169634102-039f2b99-04fb-4433-b683-e2bbe07a9cc3.png" height="50" />

Here are the proposals that included in the specification.

- JMESPath Community General Behaviour
  - **JEP-019** Standardizing [pipe expression evaluation]( https://github.com/jmespath-community/jmespath.spec/discussions/113#discussioncomment-4000862). ✔
  - **JEP-015** Enabling [string slices](https://github.com/jmespath-community/jmespath.spec/discussions/26#discussioncomment-3284127). ✔
- New Syntax
  - **JEP-016** [`arithmetic-expression`](https://github.com/jmespath-community/jmespath.spec/discussions/25#discussioncomment-3277652).✔
  - **JEP-017** [Root Node Reference](https://github.com/jmespath-community/jmespath.spec/discussions/18#discussion-3913993) with the `$` token. ✔
  - **JEP-020** [Compact syntax](https://github.com/jmespath-community/jmespath.spec/discussions/19) for `multi-select-hash`.
- New Functions
  - **JEP-011** [Lexical Scoping](https://github.com/jmespath-community/jmespath.spec/discussions/24#discussioncomment-3285710) and the `let()` function. ✔
  - **JEP-013** [Object Manipulation](https://github.com/jmespath-community/jmespath.spec/discussions/47#discussioncomment-3308897) `items()`, `from_items()` and `zip()` functions. ✔
  - **JEP-014** [String](https://github.com/jmespath-community/jmespath.spec/discussions/21#discussioncomment-3869583) functions. ✔
  - **JEP-018** [Grouping](https://github.com/jmespath-community/jmespath.spec/discussions/96#discussion-4282156) `group_by()` function. ✔

All those features [are demonstrated here](https://jmespath.site/#preview).

## History

|Date|Description|
|---|---|
|2022-10-13|Lexical Scoping proposal has been assigned number JEP-011
|2022-10-13|Object Manipulation Functions proposal has been assigned number JEP-013
|2022-10-13|String Functions proposal has been assigned number JEP-014
|2022-10-14|String Slices proposal has been assigned number JEP-015
|2022-10-14|Arithmetic Expressions proposal has been assigned number JEP-016
|2022-10-14|Root Node Reference proposal has been assigned number JEP-017
|2022-10-20|Grouping Function proposal has been assigned number JEP-018
|2022-10-29|Added proposal to standardize pipe expression evaluation
|2022-11-03|Compact Syntax for `multi-select-hash` proposal has been assigned number JEP-020
|2023-02-28|Edited post to document the exact scope of Milestone-1.1

**Comment by springcomp** [comment]
I still feel like we would need to introduce a [_reduce_ operation](https://github.com/jmespath-community/jmespath.spec/discussions/48), but we need wider community feedback. In particular, we need to contrast [my preferred suggestion](https://github.com/jmespath-community/jmespath.spec/discussions/48#discussioncomment-3314792) with the need to [include a new token](https://github.com/jmespath-community/jmespath.spec/discussions/18#discussioncomment-3322740) – _e.g_ `$` – to refer to the accumulated value within a reduce operation, or even in a wider context. Incidentally, it seems my proposal [was almost being considered](https://github.com/jmespath/jmespath.php/pull/30#issuecomment-264668696) early on in the design of JMESPath.

I also think standardizing [undocumented behaviour](https://github.com/jmespath-community/jmespath.spec/discussions/82) and enabling [empty `multi-select-hash`](https://github.com/jmespath-community/jmespath.spec/discussions/93) should be included in this milestone.

What do you think?

**Comment by springcomp** [comment]
Using `$` as a [reference to the root node](https://github.com/jmespath-community/jmespath.spec/discussions/18#discussion-3913993) is a specific case of using the `let()` function.

Some members feel that it conflicts with the [Lexical Scoping](https://github.com/jmespath-community/jmespath.spec/discussions/24#discussioncomment-3285710) proposal.
Others think it simplifies a common case.

Please, vote.
Should we include both?

**Reply by hell-racer** [reply]
I think Root Node Reference should be in New Syntax section, and not in New Functions section.

**Reply by springcomp** [reply]
Thanks for the feedback. Indeed you are right.

**Comment by kit494way** [comment]
It would be good to know in which language each specification is implemented.

**Comment by springcomp** [comment]
> It would be good to know in which language each specification is implemented.

I think the [libraries page](https://jmespath.site/main/#libraries) is trying to convey just that.
Please, let me know how I can improve it and make it clearer.
